bundles/org.openhab.binding.ecovacs/src/main/java/org/openhab/binding/ecovacs/internal/handler/EcovacsApiHandler.java

```diff
@@ -0,0 +1,174 @@
+        if (discoveryService != null) {
+            discoveryService.stopScan();
+        }
+    }
+
+    @Override
+    public Collection<Class<? extends ThingHandlerService>> getServices() {
+        return Collections.singleton(EcovacsDeviceDiscoveryService.class);
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        if (RefreshType.REFRESH == command) {
+            logger.debug("Refreshing Ecovacs API account '{}'", getThing().getUID().getId());
+            initializeApi();
+        }
+    }
+
+    public void onLoginExpired() {
+        logger.debug("Ecovacs API login for account '{}' expired, logging in again", getThing().getUID().getId());
+        final EcovacsApi api = this.api;
+        if (api != null) {
+            loginToApi(api);
+        }
+    }
+
+    private void initializeApi() {
+        String country = localeProvider.getLocale().getCountry();
+        if (country.isEmpty()) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                    "@text/offline.config-error-no-country");
+            return;
+        }
+
+        EcovacsApi api = createApi("", country);
+        loginToApi(api);
+    }
+
+    private EcovacsApi createApi(String deviceIdSuffix, String country) {
+        EcovacsApiConfiguration config = getConfigAs(EcovacsApiConfiguration.class);
+        String deviceId = config.installId + deviceIdSuffix;
+        org.openhab.binding.ecovacs.internal.api.EcovacsApiConfiguration apiConfig = new org.openhab.binding.ecovacs.internal.api.EcovacsApiConfiguration(
+                deviceId, config.email, config.password, config.continent, country, "EN", CLIENT_KEY, CLIENT_SECRET,
+                AUTH_CLIENT_KEY, AUTH_CLIENT_SECRET);
+
+        return EcovacsApi.create(httpClient, apiConfig);
+    }
+
+    private synchronized void loginToApi(final EcovacsApi api) {
+        Future<?> loginFuture = this.loginFuture;
+        if (loginFuture != null && !loginFuture.isDone()) {
+            return;
+        }
+        loginFuture = scheduler.submit(() -> {
+            try {
+                api.loginAndGetAccessToken();
+                this.api = api;
+                updateStatus(ThingStatus.ONLINE);
+
+                logger.debug("Ecovacs API initialized");
+                final EcovacsDeviceDiscoveryService discoveryService = this.discoveryService;
+                if (discoveryService != null) {
+                    discoveryService.startScan();
+                }
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                updateStatus(ThingStatus.OFFLINE);
+                this.api = null;
+            } catch (EcovacsApiException e) {
+                logger.debug("Ecovacs API login failed", e);
+                this.api = null;
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
```

Comment:
Can this happen because of network error, for instance? Will the binding be able to recover from this, i.e. automatically retry?

Reply:
Added retry. I'm not entirely sure yet it's a good idea though: One possible occasion of failure is wrong username/password. This is mentioned in the detail status message in that case, but with a retry mechanism, if left unattended, it could lead to the account being (temporarily) blocked from further login attempts. I unfortunately can't really distinguish login failures from other failures because I don't get a clear indication back (such as HTTP 401), but just a JSON object that looks like this:

```json
"{"code":"10101","msg":"Wrong password. Unable to log in. Please try again after 15 minutes.","time":1678369207194,"data":null,"success":false}"
```

The possible codes are of course undocumented :-/

It's probably not a big problem in practice, but what are your thoughts on this? Retry in any case? Retry on timeouts only (but temporary API glitches might cause permanent failure then as well)?

Reply:
Can you detect this known code (10101) and avoid retries for 15 minutes in this case? The issue could either be resolved by changing the password to match the configuration, in which case the retry itself could bring the thing online. Or, perhaps more commonly, the user will adapt the configuration, and the thing will be initialized again when saving the configuration.

Other than that, in general I think retrying will make the binding more reliable. Of course the level of sophistication can be discussed, i.e. retry with exponential backoff and jitter might be a good choice, but could also be overkill. I didn't check your retry algorithm yet, but even simple linear retry is better than no retrying.

Reply:
> Can you detect this known code (10101) and avoid retries for 15 minutes in this case?

The 15 minutes aren't a fixed value, but rather a countdown (trying 3 minutes later will yield 'Please try again after 12 minutes.'. I really don't want to parse that countdown value out of the error message ;-)
I'm also a bit reluctant to hardcode that value here, since it's just the value I happened to get when trying this once. With the general quality of their API, I'm really unsure whether it's the only possible error code for such a case.

Further attempts in the wait period don't seem to prolong it though (and don't seem to lead to blocking permanently), so it's probably not a big problem (if at all) in practice. I guess most users will check initialization right after creating a new thing (here: API bridge), so they'll likely notice the password mismatch.

Reply:
Agreed.